File: backend/permissions.py
from fastapi import Depends, HTTPException
import models
from auth import get_current_user

ADMIN = "Admin"
STOREKEEPER = "Storekeeper"
TEACHER = "Teacher"


def require_roles(*roles):
    def dep(user: models.User = Depends(get_current_user)):
        if user.role not in roles:
            raise HTTPException(403, "Permission denied")
        return user
    return dep


def require_admin(user: models.User = Depends(get_current_user)):
    if user.role != ADMIN:
        raise HTTPException(403, "Admin only")
    return user


def require_staff(user: models.User = Depends(get_current_user)):
    """Admin or Storekeeper."""
    if user.role not in (ADMIN, STOREKEEPER):
        raise HTTPException(403, "Staff only")
    return user

File: frontend/src/utils/permissions.js
// Frontend permission helpers (UI hint only — backend enforces real permissions).
export const ROLES = { ADMIN: 'Admin', STOREKEEPER: 'Storekeeper', TEACHER: 'Teacher' }

export const can = (user, action) => {
  if (!user) return false
  const role = user.role
  const map = {
    'manage_users': ['Admin'],
    'view_audit': ['Admin'],
    'manage_settings': ['Admin'],
    'approve_po': ['Admin'],
    'delete_critical': ['Admin'],
    'export_reports': ['Admin'],
    'manage_inventory': ['Admin', 'Storekeeper'],
    'manage_assignments': ['Admin', 'Storekeeper'],
    'manage_receiving': ['Admin', 'Storekeeper'],
    'manage_returns': ['Admin', 'Storekeeper'],
    'manage_po': ['Admin', 'Storekeeper'],
    'view_suppliers': ['Admin', 'Storekeeper'],
    'view_departments': ['Admin', 'Storekeeper'],
    'view_reports': ['Admin', 'Storekeeper'],
    'view_movements': ['Admin', 'Storekeeper'],
    'view_inventory': ['Admin', 'Storekeeper', 'Teacher'],
    'view_assignments': ['Admin', 'Storekeeper', 'Teacher'],
    'view_dashboard': ['Admin', 'Storekeeper', 'Teacher'],
    'edit_profile': ['Admin', 'Storekeeper', 'Teacher'],
  }
  return (map[action] || []).includes(role)
}
